### PR TITLE
docs: add formal governance system

### PR DESCRIPTION
**Pull request by DomT4**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-----

This is an attempt to find a compromise, and is built off of the discussion that happened in https://github.com/Homebrew/brew/pull/4759.

It goes further than https://github.com/Homebrew/brew/pull/4846 in formalising the powers and responsibilities of the lead maintainer and I'm sure for that reason it will be considerably controversial, but I also don't think there's anything in here that anyone should fear or feel attacked by, and it doesn't go nearly as far as abolishing the lead maintainer position. Indeed, it provides a mechanism for the lead maintainer position to regularly & formally be affirmed as desirable.

This takes a whack at comprehensively documenting the system of governance that has built up over the years, and clears up some of the powers and responsibilities exercised and expected by each tier. I very much hope to resolve this positively, and I welcome both formal all-maintainer & PLC votes on the contents of these documents if it proves necessary or desirable.

As a consequence of filing this PR I am suspending my request in https://github.com/Homebrew/brew/pull/4846#pullrequestreview-153203292, and hope to withdraw that request sooner rather than later. That request is not something I enjoyed feeling necessary, and frankly this has all become a bit of a 💩-show, which I hope we can now resolve a little less dramatically.

### CODE REVIEW

**Comment by DomT4** [comment]
> I appreciate that you have responded with a formal proposal, but this document has some big changes that are concerning to me.

Thanks for being the first to review! I'm unsure whether to sit back and let the reviews flood in & then try to handle the inevitable flood, or whether to try and more proactively handle them. For now I'll try replying as rapidly as I can and take a step back if it becomes too real-time to handle, heh.

I think it's perhaps worth me saying that I've written this document from more of a "de facto" viewpoint than a "de jure" one. For example, since I believe Mike used it in his first PR _(Apologies if I'm misremembering that!)_, in https://github.com/Homebrew/homebrew-core/pull/31445 there were seemingly two maintainers in favour of the PR being merged & one against, but Mike's vote instead of equalising things at 2-2 made it more like 3-2 & simultaneously immediately fast-forwarded to the end of overtime 🏈.

Additionally, in https://github.com/Homebrew/homebrew-core/pull/30229 there were two, potentially three, maintainers leaning in favour before Mike's comment but the chances of it being merged after dropped _hugely_. His concerns/hesitations were justified and in further discussion proven correct, but for better or worse I think if I hadn't mentioned it in Slack and asked for opinions we'd have a `libclang` formula today. So in reality, even if not in writing, the lead maintainer carries an outsized influence on the final outcome of a pull request.

Some of this is also guided by Mike's public comments in the prior PR, especially [this one](https://github.com/Homebrew/brew/pull/4759#issuecomment-416318674). That comment runs counter to what we have documented previously, so I think it's better to try and run with things as they are and work with that. To vaguely attempt to avoid overly-punchy blowback, I'll note none of this is beefing with anyone, just an attempt to look at the words/decisions said/taken and work with that reality.

File: docs/Governance.md [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+
+Regular maintainers are people who have commit access to Homebrew but are not
+expected or obligated to help with legal, technical or administrative issues
+that arise impacting the project itself. Regular maintainers can be members of
+multiple GitHub repositories or may only be maintainers of one repository, such
+as `homebrew/livecheck`.
+
+Regular maintainers spend almost all of their time around Homebrew providing
+code review or submitting pull requests of their own, although they should
+always prioritise the work of others above their own, with rare exceptions such
```
[the remark this answers is not in the document]
Comment (DomT4) [reply docs/Governance.md]: Apologies, I can reword this if multiple people find it confusing.

It's essentially trying to communicate that for example if there's 20 Issues open in core and 15 PRs waiting for review maintainers shouldn't be prioritising filing another 25 PRs; they should instead be doing code review of other PRs, looking at the Issues and seeing if they can help with any of those, etc... and then if they have time or can't help any more, etc dive into spending their time on their own priorities.

In an ideal world I'd be looking at 2-4 PRs/Issues created by other people as a minimum before filing one of my own, for example.

File: docs/Governance.md [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+### The Lead Maintainer
+
+The lead maintainer is effectively the "face of Homebrew" and the primary point
+of contact between Homebrew and external bodies, such as Bintray or the Software
+Freedom Conservancy, as well as being responsible for managing the project's
+finances. There is only ever one lead maintainer for the entire project.
+
+The lead maintainer is expected to largely direct the priorities & aims of the
+project, and consequently has significant influence over whether a suggested
+feature/change gains traction or whether a pull request is merged. It is the
```
[the remark this answers is not in the document]
Comment (DomT4) [reply docs/Governance.md]: Hopefully addressed this in https://github.com/Homebrew/brew/pull/4864#issuecomment-419693372.

File: docs/Governance.md [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+can be withdrawn at any time prior to acceptance should an issue be discovered.
+
+The process of inviting maintainers is documented
+[here](New-Maintainer-Checklist.md#new-maintainer-checklist).
+
+### Overriding The Lead Maintainer
+
+The lead maintainer has broad authority over the direction of the project and
+acceptability of suggested features or changes, as discussed earlier. This
+authority should be respected and overridden rarely. Every reasonable effort
```
[the remark this answers is not in the document]
Comment (DomT4) [reply docs/Governance.md]: Partially addressed in https://github.com/Homebrew/brew/pull/4864#issuecomment-419693372 as well, hopefully, but this section is more designed to address https://github.com/Homebrew/brew/pull/4759#issuecomment-416318674 and the concerns people (including myself) had around that.

**Comment by vitorgalvao** [comment]
The system to pick the holder of the position looks like a potential source of stress. Voting on a specific day to keep the role and decide who stays there is essentially a yearly election. It’s not unreasonable to think the lead maintainer will, even if subconsciously, start to opt for only non-controversial decisions as election day creeps in.

The lead maintainer has tough choices to make, and one of their duties is to make them for the good of the organisation. With the prospect of a nearing election, those decisions might change and not be the best for the project. Yes, ideally that should not happen, but both the lead maintainer and the voters are human; even if the lead maintainer has done a stellar work all year, decisions most recent in memory will carry a greater weight.

From what I understood, one of your concerns with the current system is people not voicing their opinion, as they may feel it’s futile when there‘s a lead maintainer that disagrees. I feel like this system will lead to the same problem, only replacing the people who experience it.

How would the process of replacing a lead maintainer go today, with the current system? I’m guessing someone would raise a concern, independently of what day it is, and discuss it within the PLC, including the lead maintainer, to try to solve the problem. If it were decided the problem was the lead maintainer themselves, then you’d go through the process of picking a new one, and so on. Seems like a fairly organised process that could be triggered on demand, rather than making it a yearly obligation. Now if the rule were “this can not happen more than once a year”, to prevent abuses to the system by a person that wanted to remove the lead maintainer, that’d make sense for me.

---

For context, I’ll say I’m fine with the system as is, with a lead maintainer and Mike in that position. But even if I don’t agree with that part of the proposal (haven’t yet finished reading the whole thing), my comments are made in the sense of trying to make the suggestions in this PR the best they can be for the health of Homebrew.

File: docs/Governance.md [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+Any maintainer or member of the PLC can nominate a new lead maintainer, but
+confirming a new lead maintainer requires every member of the PLC to vote
+and 2/3 of those votes to be in the affirmative.
+
+### Maintainer Pull Requests and Issues
+
+Maintainers may not close other maintainer's Pull Requests or Issues without
+the consent of the maintainer in question. The two exceptions of this rule are:
+
+_Inactivity_: If a Pull Request or Issue is inactive for 28 days, any maintainer
```
Comment (vitorgalvao) [inline docs/Governance.md]: Doesn’t the `stale-bot` do this already?

File: docs/Governance.md [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+is welcome to nudge the maintainer who opened the PR/Issue for an update on
+where things stand, and if one is not provided within two days the issue can be
+closed by any maintainer. If the maintainer who opened the PR/Issue wishes to
+return to it later they are welcome to reopen the PR/Issue.
+
+_Simple Changes_: If a maintainer submits, for example, a minor documentation
+tweak to `homebrew/brew` or a simple version bump to `homebrew/core` any
+maintainer is welcome to merge that in without waiting for the original
+maintainer to do so. If maintainers object to this they should add the
+`do not merge` label to the PR/Issue. That label may be ignored if the issue
```
Comment (vitorgalvao) [inline docs/Governance.md]: If the PR is “an urgent security fix or critical to the continued function of Homebrew's CI”, than it’s probably a bad idea to ignore `do not merge`. If the label is there, presumably the PR is not yet up to snuff and may break something else.

File: docs/Governance.md [hunk 6]
```diff
@@ -0,0 +1,223 @@
...
+
+Maintainers should not routinely revert the work of other maintainers without
+first discussing what the issue is with that maintainer. However, if the
+problem is significant and the original maintainer cannot be reached within an
+hour any maintainer may revert a problematic commit. It is recommended to ping
+the original maintainer into any revert PR so they can comment later if they
+wish to, and so they're aware of the revert without having to go fishing in the
+commit log.
+
+### Maintainer Conduct
```
Comment (vitorgalvao) [inline docs/Governance.md]: This whole section uses `&` instead of `and`. Is that intentional?

File: docs/Governance.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (fxcoudert) [reply docs/Governance.md]: I think I understand where you're coming at with this, but maintainers also regularly open PRs for stuff that people simply don't contribute, yet is beneficial for Homebrew. I'm thinking of maintainers who have clear expertise in an area that most others are lacking (like gtk build system, etc.).

File: docs/Governance.md [hunk 7]
```diff
@@ -0,0 +1,223 @@
...
+and nobody has unchecked power or influence. The below list covers various
+situations that may arise and how to handle them.
+
+### The Lead Maintainer Position & The Holder Of That Position
+
+At least once a year, provisionally September 14th, the PLC is mandated to
+affirm or withhold its consent for the lead maintainer role to exist in its
+current form as a distinct entity. The lead maintainer may vote on this issue.
+In the event of a tie regular maintainers are asked to vote on the issue and
+collectively act as the sole tie-breaking vote.
```
Comment (fxcoudert) [inline docs/Governance.md]: I'm not sure voting once a year on this specific item is of particular interest. Maybe a more open question is: how does the governance system described here evolve over time? This includes the lead maintainer position, but it's not restricted to that, there are more aspects that could change over time, as is seen necessary.

File: docs/Governance.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (DomT4) [reply docs/Governance.md]: Aye, it's a fine line to walk. This was mostly born out of a conversation elsewhere, and I don't feel that strongly about documenting it, so I'm fine with trimming the block if it's causing confusion or people are worried about it.

File: docs/Governance.md (hunk 4 above)
Comment (DomT4) [reply docs/Governance.md]: No, not in some cases. In `brew` for example there's a bot that automatically tags `brew` PRs with the `in progress` label, which stops stale-bot doing its thing.

File: docs/Governance.md (hunk 5 above)
Comment (DomT4) [reply docs/Governance.md]: My train of thought was something along the lines of:

1) Dave files a PR containing a critical security update for, say, `openssl`.
2) Dave tags this `do not merge` because they'd like to merge it in themselves, perhaps because they like to have all their merged commits GPG signed with their own key.
3) The CI build takes 5 hours to complete, in which time Dave has given up glaring at the build log & vanished offline. They don't return within an hour or so of the CI build finishing, and Sally decides to merge Dave's PR based on the CI result looking okay & the scale of changes being relatively minor.

I think Sally's actions would be justifiable there.

File: docs/Governance.md (hunk 6 above)
Comment (DomT4) [reply docs/Governance.md]: Not a conscious choice. I can change it it people prefer `and` throughout; I just tend to lean on `&` to condense things a little.

Comment (sjackman) [reply docs/Governance.md]: I prefer `and`

File: docs/Governance.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (fxcoudert) [reply docs/Governance.md]: I think simpler is better: if people are of good will, things will work out. If not, they will always manage to find a way to argue about the written rule anyway :)

**Comment by DomT4** [comment]
> Voting on a specific day to keep the role and decide who stays there is essentially a yearly election.

Aye, the intention there is that the role be reaffirmed once a year. I went for a specific day as a way of A) establishing that the system frowns upon people forcing votes on the role 3/4/5/6/7/8 times a year, but also B) minimises the ability of anyone (including the lead maintainer) to pick a time/date that suits them.

If Bob thinks or knows that "Lead Maintainer A" is going to have to push through something deeply controversial in July for example, and Bob thinks its time for a change of leadership, Bob could wait till the height of the controversy & then move a leadership challenge.

With elections one side always has some built-in immediate advantage, and it largely comes down to trying to minimise that or make that built-in advantage more transparent, and at least by setting a date I think it lets people be aware that if the lead maintainer suddenly starts sending everyone chocolate in the weeks leading up to the vote there's an obvious enough reason for it, whereas in the moment people may not consider Bob's thought process as much.

> How would the process of replacing a lead maintainer go today, with the current system? I’m guessing someone would raise a concern, independently of what day it is, and discuss it within the PLC, including the lead maintainer, to try to solve the problem.

Yes, a PLC member can trigger a vote. I can't get into specifics, but the current system doesn't work as well as it should or could in terms of either generating discussion or achieving any kind of outcome.

It's also quite a personal thing. You essentially have to trigger a vote of no confidence, and because this is an extraordinary event rather than a scheduled event I think there's an opening for it to be taken rather personally. I think by making it a perfectly ordinary event you make discussion easier, without it becoming such a mess, at least... I hope that makes discussion easier, because it's hard to imagine it making discussion _harder_. That's about as much as I can say, at this point.

> Now if the rule were “this can not happen more than once a year”, to prevent abuses to the system by a person that wanted to remove the lead maintainer, that’d make sense for me.

I thought about this, and I don't have an objection to this, but at the same time I also think it wise to have _some_ kind of safety net.

If the election isn't till September and in January the lead maintainer pushes through a system where we start charging people per bottle download despite the team being considerably against this and it inevitably impacting project users in a significantly negative way, and the lead maintainer does not accept an override vote, there _has_ to be some method of enforcement.

The above scenario is a completely made up one, for clarity. The intention there is to provide a last resort to the team that isn't either A) resigning in protest, B) sitting on their hands unhappily for 8 months until they can take remedial steps. It is intended to be an absolute last resort, and heavily frowned upon if people invoke it outside of the regular timetable, but the aim is to provide that safety net in the event one day Homebrew gets [McAfee'd](https://github.com/Homebrew/brew/pull/4759#issuecomment-416612398).

> my comments are made in the sense of trying to make the suggestions in this PR the best they can be for the health of Homebrew.

I appreciate your review! Thanks for taking the time 🙇.

File: docs/Governance.md (hunk 6 above)
Comment (DomT4) [reply docs/Governance.md]: Will change 👍.

File: docs/Governance.md (hunk 7 above)
Comment (DomT4) [reply docs/Governance.md]: > there are more aspects that could change over time, as is seen necessary.

I note near the top this document can be modified by any simple-majority PLC vote, which allows for it to be a kind of living document rather than something that always has to be the way that it is now. It would be a case of someone simply suggesting something to the PLC, the PLC voting on it & making any necessary changes according to that vote.

I'm not sure if that answers your question or whether I'm misunderstanding slightly?

File: docs/Governance.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (DomT4) [reply docs/Governance.md]: Removed the relevant blob.

**Comment by fxcoudert** [comment]
A few points I'd like to add to this discussion, from the point of view of someone who has been contributing to volunteer organisations and charities (and sometimes running/managing them) for 20 years now. In my opinion:

- Having an experienced "person in charge", with final say in lack of consensus on big issues or long-term projects is important. 
- Barring that, there needs to be a strong PLC or steering committee, which intervenes in both governance and strategic discussions. These discussions may be public or private, but there needs to be an active governing body, with active discussion of key points. I am not sure how active the PLC currently is (and do not know who its members are), but clearly I do not feel it has guided this debate.
- All of us are merely, day after day, dealing with Homebrew's issues as best we can, some in a limited area and some in a more comprehensive front. Some of the hypotheticals that have been thrown around are not very interesting, and I believe we should focus on helping things go better in our everyday interactions. Building strawmen is not worth the time, and if someone really wanted to screw the project, they probably could do it whatever guidelines we can write!
- I am a saddened by the fact that the situation has turned relatively personal (or, maybe not "personal", but accusatory) in a relatively short time. Some of the behaviour displayed has not been professional, and the atmosphere is made heavier by it.
- Pull requests like this one, where people comment mostly on small items, may lead the discussion to miss the broader picture.

My preferred suggestion to move forward would be to focus on making the PLC a more efficient place for debate and strategic discussion (unless it already is, but the rest of the community has no idea about it). I believe this is the most important part of this discussion, to cool down and simply find a way that we can work forward.

Finally, I'll say that in my philosophy, the role of a leader is not primarily to make rules or decisions, but to deal with all the crappy stuff that is preventing the rest of the team to work efficiently (whether it be politics, finances, communication, ops, or whatever). _I also believe the current holder of the position has been doing that quite efficiently — and thanklessly._

**Comment by vitorgalvao** [comment]
> *I also believe the current holder of the position has been doing that quite efficiently — and thanklessly.*

I wanted to touch on that point in my previous comment but forgot. Adding yet another possible source of stress to the lead maintainer is something I’d rather avoid. Whoever holds the position, I want them to enjoy working on Homebrew on some level, not add so many worries they eventually burn out and think “why am I even bothering?”. If it’s the lead maintainer’s job to ensure things run smoothly for the other maintainers, to some extent it’s said maintainers’ job to make it easy being managed. In the end, we’re all giving the same to the project — our time — and doing it for no pay, so having the most responsibility and putting in the most work translates to more trouble but not more reward.

**Comment by sjackman** [comment]
I'm in favour of having a lead maintainer position. I feel the current lead maintainer has been doing a good job.

I like both @jonchang and @fxcoudert comparison of Homebrew's governance to the governance of a not-for-profit; I serve on the board of a not-for-profit theatre company. The Project Leadership Committee (PLC) is effectively the board of Homebrew. I think of the of Lead Maintainer as the Artistic Director of an arts organization, or in more general terms, the chief executive of the organization, which is a separate role with different duties than the president of the board.

Debian's governance model and roles of Debian-Project-Leader (DPL) and the technical committee may be a good model.
https://www.debian.org/devel/leader
https://www.debian.org/devel/tech-ctte

**Comment by sjackman** [comment]
The governance model described by the bylaws of the organization ought to be a concise document. Further elaboration describing the roles can be in a separate document. For example…

The membership elects the board (PLC) at the annual general meeting (AGM).
The board appoints the executive director (lead maintainer) by resolution.
New members are admitted by the board, or possibly a membership committee.
New members of the board are typically nominated by the current board, but ultimately the membership elects the board.
The bylaws define how votes are conducted, including requirements for quorum.

Determining who comprise the membership, the board, and the executive of Homebrew are a separate topic from the technical topic of how the code base evolves. It may help the discussion to separate the two topics.

**Comment by mistydemeo** [comment]
> The governance model described by the bylaws of the organization ought to be a concise document. Further elaboration describing the roles can be in a separate document.

I think this is a good call; the overarching structure of an organization and the day-to-day processes are very different concerns, and it's worth having those in different places.

> My preferred suggestion to move forward would be to focus on making the PLC a more efficient place for debate and strategic discussion

Also agreed - I think this is important, and your point that "the rest of the community has no idea about it" is also important. Accountability from the PLC is just as important as accountability from the lead maintainer.

The Debian documents @sjackman linked do a good job of outlining their appointment model, and it sounds like one that could work us given some time to transition into it. A few takeaways for me from the Debian documents specifically:

* Debian doesn't just outline the election process, it also outlines the timeline. It sounds like they've put a good amount of thought into helping the process avoid entering a "constant election" mode, where the leader has to spend more time worrying about reelection than doing day-to-day work. Given the substantially smaller size of Homebrew as a project, Debian's campaigning process would be overkill for us, but it's useful to learn from.
* Debian's Technical Committee's decisions are outlined on the page that also defines their membership. It looks like this helps them remain accountable - if their major votes are published in this way, it becomes clear what they do.
* I love point six on the technical committee's "how to refer a question" list.

**Comment by scpeters** [comment]
> > I appreciate that you have responded with a formal proposal, but this document has some big changes that are concerning to me.

> I think it's perhaps worth me saying that I've written this document from more of a "de facto" viewpoint than a "de jure" one.

Yeah, it was a bit confusing to read this document, which seems in conflict with one of your [other comments](https://github.com/Homebrew/brew/pull/4759#issuecomment-416328532), but "de facto" vs. "de jure" helps explain the difference.

There is some value in "de facto" assessment of where we are and where we have been and being honest about it, but I would hope that a governance model could include more "de jure", or how we would like the project to be run.

**Comment by scpeters** [comment]
> @fxcoudert wrote:
> * Having an experienced "person in charge", with final say in lack of consensus on big issues or long-term projects is important.
> * Barring that, there needs to be a strong PLC or steering committee, which intervenes in both governance and strategic discussions. These discussions may be public or private, but there needs to be an active governing body, with active discussion of key points. I am not sure how active the PLC currently is (and do not know who its members are), but clearly I do not feel it has guided this debate.

I agree with both of these points. I made some comments above where I perceived a devaluing of maintainer consensus from the changes in this PR, but we can't always reach consensus, and a leader who listens and helps guide a path forward is immensely valuable.

**Comment by DomT4** [comment]
You are welcome to use this PR as a basis for future discussion, but the issue has been forced and my resignation has been mandated by Mike, which will be delivered imminently.

**Comment by fxcoudert** [comment]
Posting.

**Comment by fxcoudert** [comment]
Posting this here on behalf of @MikeMcQuaid, for reasons explained below

----

Hi 👋  ! Mike McQuaid, current, elected, Homebrew Project Leader.

DomT4's last comment on this thread was:

> You are welcome to use this PR as a basis for future discussion, but the issue has been forced and my resignation has been mandated by Mike, which will be delivered immediately.

I had edited this, with consultation with other Homebrew maintainers, to:

> You are welcome to use this PR as a basis for future discussion and my resignation will be delivered immediately.

This was not the right decision for me to make. I had attempted to revert the edit but by that point I had been blocked by DomT4 which prevented me from editing his post or posting on this thread. As a result, I deleted the comment as I felt that was more fair to DomT4 than maintaining an inaccurate comment.

This was also not the right decision to make. I am unable to undo either of the above actions or post on this thread (hence this being posted by another user) due to GitHub’s moderation system.

I would like to publicly apologise to DomT4 and the Homebrew community for my mistakes in this thread and I will do better in future.

On DomT4's resignation specifically, he said he was already leaving the project, changed his mind and, after consultation and agreement from other maintainers, I removed him from the project. Several maintainers had said they would leave the Homebrew project if he did not. There was other, personal factors that were involved in this decision which I’m unwilling to disclose as it would violate the privacy of the individuals involved. Regardless, DomT4 did feel he was forced out of the project and his comments to that effect should have remained here.

Since this pull request was closed, [I announced a date and stood down as Lead Maintainer](https://github.com/Homebrew/brew/pull/4884) and the Homebrew maintainers met in person and came up with [a public governance document](https://github.com/Homebrew/brew/pull/5739). This resulted in me being elected unanimously-bar-one as Project Leader, PLC and TSC members being elected. I am the current, 2019-2020 Project Leader, on the PLC and TSC but this time next year I will be able to be in a maximum of two roles and will have had to stand for election again.

In summary, this PR was part of a conversation that resulted in meaningful changes to my role and Homebrew’s current and future governance. Unfortunately, the way the conversation was handled was not ideal and mistakes were made but I and Homebrew will (as we always try to do) learn, iterate and improve in future.

Thanks for using Homebrew!
